File: other/merge_intervals.py
# 给出一组区间，请合并所有重叠的区间。
# input e.g. [[10,30],[20,60],[80,100],[150,180]]
# output e.g. [[10,60],[80,100],[150,180]]

#
# @param intervals Interval类一维数组 
# @return Interval类一维数组
#

class Interval:
    def __init__(self, a=0, b=0):
        self.start = a
        self.end = b

class Solution:
    def merge(self , intervals):
        # write code here
        if len(intervals) <= 1:
            return intervals

        intervals.sort(key = lambda x: x.start)

        merged = []
        for i in intervals:
            if len(merged) == 0 or i.start > merged[-1].end:
                merged.append(i)
            elif i.start <= merged[-1].end:
                if i.end <= merged[-1].end:
                    continue
                else:
                    merged[-1].end = i.end
        return merged

# Test
interval1 = Interval(50, 60)
interval2 = Interval(20, 60)
interval3 = Interval(50, 100)
interval4 = Interval(80, 120)
solution = Solution()

print(solution.merge([interval1, interval2, interval3, interval4])[0].start)
print(solution.merge([interval1, interval2, interval3, interval4])[0].end)       

File: other/merge_intervals.cpp
// leetcode 56. mergeIntervals cpp

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) 
    {
        int n = intervals.size();
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> ret;
        int cur = 0;
        while(cur < n)
        {
            int max_val = intervals[cur][1];

            if(cur == n - 1)
            {
                ret.push_back(intervals[cur]);
                break;
            }
            for(int seek = cur + 1; seek < n; seek++)
            {
                if(max_val >= intervals[seek][0])
                {
                    if(seek == n - 1)
                    {
                        max_val = max(max_val, intervals[seek][1]);
                        ret.push_back({intervals[cur][0], max_val});
                        cur = n;
                    }
                    max_val = max(max_val, intervals[seek][1]);
                    continue;
                }
                else
                {
                    ret.push_back({intervals[cur][0], max_val});
                    cur = seek;
                    break;
                }
            }
        }
        return ret;
    }
};
